### PR TITLE
protection for click behind drag

### PR DESCRIPTION
**Pull request by starowo**
Adding protection for click notes behind drag notes.

And modified near note judgments.

### CODE REVIEW

File: prpr/src/judge.rs [hunk 1]
```diff
@@ -447,7 +447,7 @@ impl Judge {
                 let Some(pos) = pos[id] else { continue; };
                 for id in &idx[*st..] {
                     let note = &mut line.notes[*id as usize];
-                    if !matches!(note.judge, JudgeStatus::NotJudged) {
+                    if !matches!(note.judge, JudgeStatus::NotJudged | JudgeStatus::PreJudge) {
```
Comment (Mivik) [inline prpr/src/judge.rs:450]: reasonable

File: prpr/src/judge.rs [hunk 2]
```diff
@@ -472,13 +472,12 @@ impl Judge {
                     {
                         continue;
                     }
-                    let dt = dt
-                        + if matches!(note.kind, NoteKind::Flick | NoteKind::Drag) {
-                            0.05 /* TODO tweak*/
+                    let dt = if matches!(note.kind, NoteKind::Flick | NoteKind::Drag) {
+                            if dt <= LIMIT_PERFECT {dt.max(LIMIT_PERFECT)} else {dt.max(LIMIT_GOOD)}
```
Comment (Mivik) [inline prpr/src/judge.rs]: Could you please elaborate on this?

Comment (starowo) [reply prpr/src/judge.rs]: modified, nvm that

Comment (starowo) [reply prpr/src/judge.rs]: It's a failed conjecture of the original mechanisms.

File: prpr/src/judge.rs [hunk 3]
```diff
@@ -472,13 +472,8 @@ impl Judge {
...
                     }
-                    let dt = dt
-                        + if matches!(note.kind, NoteKind::Flick | NoteKind::Drag) {
-                            0.05 /* TODO tweak*/
-                        } else {
-                            0.
-                        };
-                    if dt + (dist / res.note_width - 1.).max(0.) * DIST_FACTOR < closest.2 + (closest.1 / res.note_width - 1.).max(0.) * DIST_FACTOR {
+                    let dt = if matches!(note.kind, NoteKind::Flick | NoteKind::Drag) {dt.max(LIMIT_PERFECT)} else {dt};
+                    if dt + (dist / res.note_width - 1.).max(0.) * DIST_FACTOR < closest.2 - 0.01 + (closest.1 / res.note_width - 1.).max(0.) * DIST_FACTOR {
```
Comment (Mivik) [inline prpr/src/judge.rs]: 0.01 here needs more tests, I'll resolve this later

Comment (starowo) [reply prpr/src/judge.rs]: I've tested a main situation.
At least for this one:
image
it works well.

Comment (starowo) [reply prpr/src/judge.rs]: Btw I dont think 10ms has an important impact. There are 20ms intervals even between two 48th notes in 180 bpm.

**Comment by Mivik** [comment]
lgtm

**Pull request merged**